src/Context/InAppPurchase/HasMatchingDomain.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopware\App\SDK\Context\InAppPurchase;
+
+use Lcobucci\JWT\Token;
+use Lcobucci\JWT\UnencryptedToken;
+use Lcobucci\JWT\Validation\Constraint;
+use Lcobucci\JWT\Validation\ConstraintViolation;
+use Shopware\App\SDK\Shop\ShopInterface;
+
+/**
+ * @phpstan-import-type InAppPurchaseArray from InAppPurchaseProvider
+ */
+class HasMatchingDomain implements Constraint
+{
+    public function __construct(private readonly ShopInterface $shop)
+    {
+    }
+
+    public function assert(Token $token): void
+    {
+        if (!$token instanceof UnencryptedToken) {
+            throw new \Exception('Incorrect token type');
+        }
+
+        /** @var InAppPurchaseArray $inAppPurchase */
+        foreach ($token->claims()->all() as $inAppPurchase) {
+            if (!\array_key_exists('sub', $inAppPurchase)) {
+                throw ConstraintViolation::error('Missing sub claim', $this);
+            }
+
+            $host = \parse_url($this->shop->getShopUrl(), \PHP_URL_HOST);
+
+            if ($inAppPurchase['sub'] !== $host) {
+                throw ConstraintViolation::error('Token domain invalid: ' . $inAppPurchase['sub'] . ', expected: ' . $host, $this);
```

Comment:
You need a valid license domain set in your shop, this should not be an issue 🤔

Reply:
I'm not sure how it would prevent local development. The token needs to have a matching domain. It's part of our validation check. I believe this should be part of the apps validation check too.

Could you maybe expand on what your concerns are here, @cyl3x ?

Reply:
I think the check is too late here. The domain check should happen in platform and if you're throwing an exception here or not doesn't matter as the app server already received the IAP JWT. So an app server can decode the JWT anyway, there is no need to pseudo-prevent this here :thinking:

Reply:
But then what is the point of sending a JWT token to the app, if it's already been validated in platform? Wasn't the whole point of having a JWT to give the app a chance to validate the purchase as well?

Reply:
The JWT is there so App Servers can rely that the IAPs they receive are validated by Shopware's internal SBP and the purchase is legit, as the token was signed by the SBP. 

The App Server should check this as well to be sure, that the message was not tampered with on the was.

Reply:
Yeah I agree, that's why I'm asking the question to @cyl3x why we are sending a JWT to the app, if it's already validated by platform.

Reply:
Because the validation could be circumvented through plugins or other code manipulation on a shop instance level. 

The App Server *must* check the JWT on its own with the public key of Shopware's SBP to be 100% certain.